Review Pull Request

WilsonSamuel02 opened the pull request.

WilsonSamuel02 committed:
Transferred project files/testing.

WilsonSamuel02 committed:
Review Pull Request

HarryJWesson commented on Game/game.cs:
```diff
@@ -0,0 +1,192 @@
+                Console.WriteLine();
+                string? userInput = Console.ReadLine();
+
+                if (userInput != null){
+                    if (userInput.ToLower() == "observe"){
+                        startRoom.GetDescription();
+                        break;
+                    }
+                    else if (userInput.ToLower() == "inventory"){
+                        player.AccessInventory();
+                    }
+                    else if (userInput.ToLower() == "stats"){
+                        player.AccessStats();
+                    }
+                    else if (userInput.ToLower() == "leave"){
+                        Console.WriteLine("You decided to leave and live another day!");
+                        gameState = (int)States.EXIT_MENU;
+
+                        return gameState;
+                    }
+                    else{
+                        Console.WriteLine("That is not an option!");
+                        Console.WriteLine("Please try again!");
+                    }
+                }
+                else{
+                    Console.WriteLine("That is not an option!");
+                    Console.WriteLine("Please try again!");
+                }
```
Could swap to a switch case statement

HarryJWesson commented on Game/game.cs:
```diff
@@ -0,0 +1,192 @@
+            int gameState;
+
+            Console.WriteLine("Blah blah blah, story here, later.");
+            Console.WriteLine("What would you like to name your warrior?");
+            string? playerName = Console.ReadLine();
+
+            Player player = new Player();
+            
+            if (playerName != null){
+                player.Name = playerName;
+            }
+            
```
Could introduce something here in case the user enters an empty string

HarryJWesson reviewed (commented):
Overall very good and very readable.
Consistent and clear formatting, helpful comments but could be more consistent throughout the code, could have more commits but the commit messages are clear and show workflow was followed locally. Good use of encapsulation with getter setters.
Lots of else if statements that could be swapped to switch case statements for visual clarity.

AlexGilmore00 commented on Program.cs:
```diff
@@ -10,11 +10,8 @@ internal class Program
-            Console.WriteLine("Waiting for your Implementation");
-            Console.WriteLine("Press any key to exit...");
-            Console.ReadKey();
+            GameLoop gameLoop = new GameLoop();
+            gameLoop.Start();
```
try catch block around game initialisation and start method call could help catch any unhandled exceptions

AlexGilmore00 commented on Entities/entity.cs:
```diff
@@ -0,0 +1,45 @@
+        private int _attack;
+        private int _defense;
+
+        public string Name{
+            get {return _name;}
+            set { _name = value;}
+        }
```
public fields like this with accessors that that just return and set the value straight without any logic can just be replaced with...
public string Name { get; private set; }
or...
public string Name { get; protected set; }
this would also allow you to remove the declaration of the private string _name variable as using an accessor like this implicitly creates that variable under the hood. this can help increase readability.
in this case, you would want to use...
public string Name { get; protected set; }
so objects that are a child of this class can set its value.

AlexGilmore00 commented on Entities/entity.cs:
```diff
@@ -0,0 +1,45 @@
+            set { _name = value;}
+        }
+
+        public int Health{
+            get {return _health;}
+            set { 
+                if (_health <= 0){ _health = 0};
+                else if (Health > _maxHealth){ _health = maxHealth};
+                else{ _health = value};
+
+            }
+        }
```
here you might want to use protected set in the accessor rather than just set.
currently Health (and therefore the private _health field) can be set by any class.
changing it to use protected set would mean it can only be set by this class and its children.

AlexGilmore00 reviewed (commented):
Overall really good.
The use of the menu class and gameloop class make the code very readable.
The use of inheritance from a parent entity is good for maintainability as well.
The way the game class is structured feels very bloated and limiting. i believe some of the while loops such as the one could be better off as methods in the menu class that can be called. this would allow you to flow more freely between these loops which would increase maintainability and scalability over the more linear approach being used now.
Many of the large if else if blocks could be replaced with switch case blocks to slightly improve time efficiency.
Addition of a test class and the use of Debug.Assert could help with testing implementation.
Could consider using classes for items as well in the future if you aren't already.

WilsonSamuel02 committed:
Fixed code errors found

WilsonSamuel02 committed:
Updated code with review comments

Certain features requested by the review comments were unable to be implemented as switch statements in c# are unable to handle incorrect input.

WilsonSamuel02 committed:
started adding xml comments on necessary methods and classes

WilsonSamuel02 committed:
add combat class testing

WilsonSamuel02 committed:
Program fixed

WilsonSamuel02 committed:
added combat class

WilsonSamuel02 committed:
Added enemy class, and various species to utilize/randomise

WilsonSamuel02 committed:
Added Boss stats and organised monster species

WilsonSamuel02 committed:
started combat system

WilsonSamuel02 committed:
added more details to player creation, such as species stats and new description menu

WilsonSamuel02 committed:
Organised file system and started on prefab system

WilsonSamuel02 committed:
Added basic starter weapons and started adding certain character creation cantrips / spells.

WilsonSamuel02 committed:
Added cantrips

WilsonSamuel02 committed:
added an introductory story and adjusted some settings

WilsonSamuel02 committed:
Merged old with new and fixed some merge issues

WilsonSamuel02 committed:
Fixed merge issues

WilsonSamuel02 committed:
Fixed boss entities and added more story elements

WilsonSamuel02 committed:
Started to add the merchant code in preperation for merchant testing, before adding th eother NPCs

WilsonSamuel02 committed:
Started creating map generator

WilsonSamuel02 committed:
started a predetermined map as a backup

WilsonSamuel02 committed:
continued dungeon generator

WilsonSamuel02 committed:
started combat system and added new enemies.

WilsonSamuel02 committed:
Added an attack program

WilsonSamuel02 committed:
Finished combat system
